examples/topology/generateSimplicityTables3D.cpp

```diff
@@ -30,96 +30,52 @@
-*/
-template <typename Object, typename Map>
-void 
-generateSimplicityTable( const typename Object::DigitalTopology & dt,
-			 Map & map )
+int main( int argc, char** argv )
 {
-  typedef typename Object::DigitalSet DigitalSet;
-  typedef typename Object::Point Point;
-  typedef typename DigitalSet::Domain Domain;
-  typedef typename Domain::ConstIterator DomainConstIterator;
-
-  Point p1 = Point::diagonal( -1 );
-  Point p2 = Point::diagonal(  1 );
-  Point c = Point::diagonal( 0 );
-  Domain domain( p1, p2 );
-  DigitalSet shapeSet( domain );
-  Object shape( dt, shapeSet );
-  unsigned int k = 0;
-  for ( DomainConstIterator it = domain.begin(); it != domain.end(); ++it )
-    if ( *it != c ) ++k;
-  ASSERT( ( k < 32 )
-	  && "[generateSimplicityTable] number of configurations is too high." );
-  unsigned int nbCfg = 1 << k;
-  for ( unsigned int cfg = 0; cfg < nbCfg; ++cfg )
-    {
-      if ( ( cfg % 1000 ) == 0 )
-	{
-	  trace.progressBar( (double) cfg, (double) nbCfg );
-	}
-      shape.pointSet().clear();
-      shape.pointSet().insert( c );
-      unsigned int mask = 1;
-      for ( DomainConstIterator it = domain.begin(); it != domain.end(); ++it )
-	{
-	  if ( *it != c )
-	    {
-	      if ( cfg & mask ) shape.pointSet().insert( *it );
-	      mask <<= 1;
-	    }
-	}
-      bool simple = shape.isSimple( c );
-      map[ cfg ] = simple;
-    }
-}
-
-
-
-int main( int /*argc*/, char** /*argv*/ )
-{
-  typedef std::bitset<67108864> ConfigMap; // 2^26
-
   using namespace Z3i;
-  trace.beginBlock ( "Generate 3d table for 26_6 topology" );
+  string error_message(
+      "Select ForegroundAdjacency for object topology:\n"
+      "- 26_6 \n"
+      "- 18_6 \n"
+      "- 6_18 \n"
+      "- 6_26 \n");
+  if (argc != 2 ){
+    cout << error_message << std::endl;
+    return 1;
+  }
+  std::string input_str = std::string(argv[1]);
+
+  using ConfigMap = std::bitset<67108864> ; // 2^26
   // Too big for stack. Use heap instead.
-  ConfigMap* table26_6 = new ConfigMap;
-  generateSimplicityTable< Object6_26 >( dt6_26, *table26_6 );
+  unique_ptr<ConfigMap> table(new ConfigMap);
+  trace.beginBlock ( "Generate 3d table for " + input_str + " topology" );
+
+  if (input_str == "26_6")
+    functions::generateSimplicityTable< Object26_6, ConfigMap >( dt26_6, *table );
+  else if (input_str == "18_6")
+    functions::generateSimplicityTable< Object18_6, ConfigMap >( dt18_6, *table );
+  else if (input_str == "6_18")
+    functions::generateSimplicityTable< Object6_18, ConfigMap >( dt6_18, *table );
+  else if (input_str == "6_26")
+    functions::generateSimplicityTable< Object6_26, ConfigMap >( dt6_26, *table );
+  else {
+    cout << error_message << endl;
+    return 1;
+  }
   trace.endBlock();
 
-  ofstream file26_6( "simplicity_table6_26.txt" );
-  file26_6 << *table26_6;
-  file26_6.close();
+  string filename = "simplicity_table" + input_str + ".txt";
+  ofstream file( filename  );
+  file << *table;
+  file.close();
```

Comment:
By the way, did you check on some real examples that my function Object::isSimple is correct for all topology pairs ? I had problem at some point (which I tried to correct at best) but I never completely check these methods. Well, they were able to thin homotopically reasonnable size examples, and for 6_18 and 6_26, they produce thick skeleton, while being thinner for other adjacencies.